prusti-viper/src/encoder/counterexample.rs

```diff
@@ -0,0 +1,184 @@
+                    } else {
+                        let note = format!("counterexample for \"{0}\"\n{0} <- {1}", place.0, entry);
+                        prusti_error.add_note(note, Some(place.1.clone()));
+                    }
+                }
+                let result_note = format!{"result <- {}", result};
+                prusti_error.add_note(result_note, None);
+            } else {
+                for (place, entry) in args {
+                    let note = format!("counterexample for \"{0}\"\n{0} <- {1}", 
+                            place.0,   
+                            entry,
+                        );
+                    prusti_error.add_note(note, Some(place.1.clone()));
+                }
+                // Todo: find span of return type to give this note a span
+                let result_note = format!("result <- {}", result);
+                prusti_error.add_note(result_note, None);
+            }
+        } 
+    }
+}
+
+impl fmt::Display for Counterexample {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Counterexample::Success{result, args, entries, is_pure} => {
+                write!(f, "Counterexample:\n");
+                write!(f, "initial args:\n");
+                for (place, entry) in args {
+                    let s = format!("{} <- {}\n", place.0, entry);
+                    write!(f, "{}", indent(s));
+                }
+                write!(f, "\nlocal values when failing:\n");
+                for (place, entry) in entries {
+                    let s = format!("{} <- {}\n", place.0, entry);
+                    write!(f, "{}", indent(s));
+                }
+                write!(f, "\nresult <- {}", result)
+            },
+            _ => write!(f, "Counterexample generation failed")
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum Entry{
+    IntEntry{value: i64},
+    BoolEntry{value: bool},
+    CharEntry{value: char},
+    RefEntry{el: Box<Entry>},
+    Struct{
+        name: String,
+        field_names: Vec<String>,
+        field_entries: Vec<Entry>
+    },
+    Enum{
+        super_name: String,
+        name: String,
+        named_fields: bool,
+        field_names: Vec<String>,
+        field_entries: Vec<Entry>
+        //note: if fields are not named, their order is important!
+        //that is why no HashMap is used
+    },
+    Tuple{
+        fields: Vec<Entry>
+    },
+    Unit,
+    UnknownEntry
+}
+
+impl fmt::Display for Entry{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self{
+            Entry::IntEntry { value } => write!(f, "{}", value),
+            Entry::BoolEntry { value } => write!(f, "{}", value),
+            Entry::CharEntry { value } => write!(f, "'{}' ({:x})", value, *value as i32),
+            Entry::RefEntry { el } => write!(f, "ref({})", indent(el.to_string())),
+            Entry::Enum { super_name, name, named_fields, field_names, field_entries } => {
```

Comment:
You can probably use [`f.debug_struct`](https://doc.rust-lang.org/std/fmt/struct.Formatter.html#method.debug_struct) (and similar) for enums, structs, and tuples, then get rid of `indent`.

Reply:
I tried this, and there is one drawback, it prints everything on one line. I guess we can discuss this, I can also try to clean up the code of my formatting. But having the fields on multiple lines with indentations is pretty useful in my opinion, especially when there are nested structures.